components/camel-minio/src/main/java/org/apache/camel/component/minio/MinioConsumer.java

```diff
@@ -0,0 +1,393 @@
+ */
+package org.apache.camel.component.minio;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.*;
+
+import io.minio.DownloadObjectArgs;
+import io.minio.MinioClient;
+import io.minio.Result;
+import io.minio.errors.InvalidBucketNameException;
+import io.minio.errors.MinioException;
+import io.minio.messages.Bucket;
+import io.minio.messages.Item;
+import org.apache.camel.AsyncCallback;
+import org.apache.camel.Exchange;
+import org.apache.camel.ExtendedExchange;
+import org.apache.camel.NoFactoryAvailableException;
+import org.apache.camel.Processor;
+import org.apache.camel.spi.Synchronization;
+import org.apache.camel.support.ScheduledBatchPollingConsumer;
+import org.apache.camel.util.CastUtils;
+import org.apache.camel.util.ObjectHelper;
+import org.apache.camel.util.URISupport;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A Consumer of messages from the Minio Storage Service.
+ */
+public class MinioConsumer extends ScheduledBatchPollingConsumer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MinioConsumer.class);
+
+    private Iterator<Result<Item>> marker;
+    private transient String minioConsumerToString;
+
+    public MinioConsumer(MinioEndpoint endpoint, Processor processor) throws NoFactoryAvailableException {
+        super(endpoint, processor);
+    }
+
+    @Override
+    protected int poll() throws Exception {
+        // must reset for each poll
+        shutdownRunningTask = null;
+        pendingExchanges = 0;
+
+        assert getConfiguration().getBucketName() != null;
+        String bucketName = getConfiguration().getBucketName();
+        MinioClient minioClient = getMinioClient();
+        String objectName = getConfiguration().getObjectName();
+        InputStream minioObject = null;
+        Queue<Exchange> exchanges = null;
+
+        if (bucketExists(minioClient, bucketName)) {
+            LOG.trace("Bucket {} exists", bucketName);
+        } else {
+            throw new InvalidBucketNameException("Bucket {} does not exists", bucketName);
+        }
+
+        if (objectName != null) {
+            LOG.trace("Getting object in bucket {} with object name {}...", bucketName, objectName);
+
+            try {
+                minioObject = getObject(bucketName, minioClient, objectName);
+                if (minioObject != null) {
+                    exchanges = createExchanges(minioObject, objectName);
+                }
+
+            } catch (Throwable e) {
+                LOG.warn("Failed to get object in bucket {} with object name {}, Error message {}", bucketName, objectName, e.getMessage());
+                throw e;
+
+            } finally {
+                //must be closed after use to release network resources.
+                try {
+                    assert minioObject != null;
+                    minioObject.close();
```

Comment:
Maybe [try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) could be used here. It would be nice to reduce nesting.